designs/2019-config-simplification/README.md

```diff
@@ -509,6 +509,25 @@ module.exports = [
 
 When ESLint uses this config, it will check each `files` pattern to determine which configs apply. Any config with a `files` pattern matching the file to lint will be extracted and used (if multiple configs match, then those configs are merged to determine the final config to use). In this way, returning an array acts exactly the same as the array in `overrides`.
 
+#### Using AND patterns for files
+
+If any entry in the `files` key is an array, then all of the patterns must match in order for a filename to be considered a match. For example:
+
+```js
+module.exports = [
+    {
+        files: [ ["*.test.*", "*.js"] ],
+        rules: {
+            semi: ["error", "always"]
+        }
+    }
```

Comment:
How can I apply a shareable config to only specific files?

Reply:
As mentioned previously, for the time being, we need to do this by documenting that shareable configs that want to be applied to multiple file patterns should be published without a `files` field specified. That way, consumers can just do something like this:

```js
const sharedConfig = require("eslint-config-shared");

module.exports = [
    {
        files: ["*.js"],
        ....sharedConfig
    },
    {
        files: ["*.mjs"],
        ...sharedConfig
    }
];
```

Because shared configs don't need to follow a specific pattern, there are any number of ways developers can create shared configs to allow end users to easily mix-and-match. If we see common patterns emerging, we can always revisit and provide some utilities to help.

Reply:
By this RFC, config can be an array, a function, or an object. We cannot use spread property syntax to merge those.

Reply:
Again, we can solve this with documentation in the short-term. No shareable config is an array or a function right now, so we can tell people that if they want to make their config usable for any file pattern, make sure it's either an object or keep it in `eslintrc` format.

Once we get some feedback and see how people are using the new config system, we can figure out what other utilities we should add to make people's lives easier.

Reply:
> No shareable config is an array or a function right now

That's not practical because most shareable configs are an array to use `extends`. It's pretty common use-case that shareable configs extend other shareable configs and plugin's recommended config.